pkg/cloud/openstack/clients/machineservice.go

```diff
@@ -357,7 +357,35 @@ func isDuplicate(list []string, name string) bool {
 	return false
 }
 
-func GetSecurityGroups(is *InstanceService, sg_param []openstackconfigv1.SecurityGroupParam) ([]string, error) {
+func GetSecurityGroups(is *InstanceService, sg_param []openstackconfigv1.SecurityGroupParam, cluster *clusterv1.Cluster, machine *clusterv1.Machine) ([]string, error) {
+	clusterName := fmt.Sprintf("%s-%s", cluster.Namespace, cluster.Name)
+	secGroupService, err := NewSecGroupService(is.networkClient)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create security group service: %v", err)
+	}
+
+	clusterStatus, err := openstackconfigv1.ClusterStatusFromProviderStatus(cluster.Status.ProviderStatus)
+	if err != nil {
+		return nil, fmt.Errorf("failed to load cluster provider status: %v", err)
+	}
+
+	clusterSpec, err := openstackconfigv1.ClusterSpecFromProviderSpec(cluster.Spec.ProviderSpec)
+	if err != nil {
+		return nil, fmt.Errorf("failed to load cluster provider spec: %v", err)
+	}
+
+	err = secGroupService.Reconcile(clusterName, *clusterSpec, clusterStatus)
+	if err != nil {
+		return nil, fmt.Errorf("failed to reconcile security groups: %v", err)
+	}
+
+	if clusterSpec.ManagedSecurityGroups {
+		if util.IsControlPlaneMachine(machine) {
```

Comment:
I understand we want to return cluster defined sec and how about add a log in order to let operator understand we ignored the sec group we defined in machine spec to avoid confusion?

Reply:
@jichenjc Sorry for the delay, had a vacation. I agree about the fact that user should be aware of ignored security groups. I updated changes, please review it again.